core/threaded/watchdog.c

```diff
@@ -2,82 +2,132 @@
  * @file
  * @author Benjamin Asch
  * @author Edward A. Lee
+ * @author Erling Jellum
  * @copyright (c) 2023, The University of California at Berkeley.
  * License: <a href="https://github.com/lf-lang/reactor-c/blob/main/LICENSE.md">BSD 2-clause</a>
  * @brief Definitions for watchdogs.
  */
 
 #include <assert.h>
 #include "watchdog.h"
-
-extern int _lf_watchdog_count;
-extern watchdog_t* _lf_watchdogs;
+#include "environment.h"
+#include "util.h"
 
 /**
  * @brief Initialize watchdog mutexes.
  * For any reactor with one or more watchdogs, the self struct should have a non-NULL
  * `reactor_mutex` field which points to an instance of `lf_mutex_t`.
- * This function initializes those mutexes.
+ * This function initializes those mutexes. It also initializes the condition
+ * variable which enables the safe termination of a running watchdog.
  */
-void _lf_initialize_watchdog_mutexes() {
-    for (int i = 0; i < _lf_watchdog_count; i++) {
-        self_base_t* current_base = _lf_watchdogs[i].base;
-        if (current_base->reactor_mutex != NULL) {
-            lf_mutex_init((lf_mutex_t*)(current_base->reactor_mutex));
+void _lf_initialize_watchdog_mutexes(environment_t *env) {
+    int ret;            
+    for (int i = 0; i < env->watchdogs_size; i++) {
+        watchdog_t *watchdog = env->watchdogs[i];
+        if (watchdog->base->reactor_mutex != NULL) {
+            ret = lf_mutex_init((lf_mutex_t*)(watchdog->base->reactor_mutex));
+            LF_ASSERTN(ret, "lf_mutex_init failed");
+        }
+        ret = lf_cond_init(&watchdog->cond, watchdog->base->reactor_mutex);
+        LF_ASSERTN(ret, "lf_cond_init failed");
+    }
+}
+
+/**
+ * @brief Terminate all watchdog threads. 
+ */
+void _lf_watchdog_terminate(environment_t *env) {
+    void *thread_return;
+    int ret;
+    for (int i = 0; i < env->watchdogs_size; i++) {
+        watchdog_t *watchdog = env->watchdogs[i];
+        if (watchdog->thread_active) {
+            ret = lf_mutex_lock(watchdog->base->reactor_mutex);
+            LF_ASSERTN(ret, "lf_mutex_lock failed");
+            lf_watchdog_stop(watchdog);
+            ret = lf_mutex_unlock(watchdog->base->reactor_mutex);
+            LF_ASSERTN(ret, "lf_mutex_unlock failed");
+            ret = lf_thread_join(watchdog->thread_id, &thread_return);
+            LF_ASSERTN(ret, "lf_thread_join failed");
         }
     }
 }
 
 /**
  * @brief Thread function for watchdog.
```

Comment:
An `@brief` tag needs a blank line below it or the rest of the paragraph is interpreted as part of the brief comment by doxygen.